File: cp_cpp/src/bin/bracketmatching.cpp
// Created by Ayush Biswas at 2026/04/17 20:37
// https://open.kattis.com/problems/bracketmatching

/// @head begin
#include <iostream>
#include <stack>
#include <map>
/// @head end

/// @code begin
using namespace std;

void solve(const int n, const string& s) {
    map<char, char> closed = {
        {')', '('},
        {']', '['},
        {'}', '{'}
    };

    stack<char> st;
    for (auto c : s) {
        if (closed.contains(c)) {
            if (st.empty() || st.top() != closed[c]) {
                cout << "Invalid";
                return;
            }
            st.pop();
        } else {
            st.push(c);
        }
    }
    if (st.empty()) {
        cout << "Valid";
    } else {
        cout << "Invalid";
    }
}

int main() {
    cin.tie(nullptr);

    int n;
    cin >> n;
    string s;
    cin >> s;

    solve(n, s);
    return 0;
}
/// @code end

File: cp_rust/src/bin/bracketmatching.rs
// Created by Ayush Biswas at 2026/05/06 15:56
// https://open.kattis.com/problems/bracketmatching
use cp_lib::*;

// @code begin
use cpio::*;
use std::collections::HashMap;
// const INF: u32 = 10u32.pow(9);
// const INF: usize = 4 * 10usize.pow(18);

sol! {
    fn solution(
        n: usize,
        s: String
    ) -> String {
        let mut stack: Vec<char> = Vec::with_capacity(n + 1);
        let opening: HashMap<char, char> = HashMap::from([
            (')', '('),
            (']', '['),
            ('}', '{')
        ]);
        for c in s.chars() {
            if opening.contains_key(&c) {
                if let Some(c_nemesis) = stack.pop() {
                   if c_nemesis != opening[&c] {
                        return "Invalid".into()
                   }
                } else {
                    return "Invalid".into()
                }
            } else {
                stack.push(c);
            }
        }
        if stack.is_empty() {
            "Valid".into()
        } else {
            "Invalid".into()
        }
    }
}

// @code end
